Add Repository Security Advisories APIs

anishrajan25 opened the pull request:
Relates to #2866 

Integrates the following 2 APIs:

- [Lists repository security advisories for an organization - /orgs/{org}/security-advisories](https://docs.github.com/en/rest/security-advisories/repository-advisories?apiVersion=2022-11-28#list-repository-security-advisories-for-an-organization)
- [Lists security advisories in a repository - /repos/{owner}/{repo}/security-advisories](https://docs.github.com/en/rest/security-advisories/repository-advisories?apiVersion=2022-11-28#list-repository-security-advisories)

anishrajan25 commented:
@gmlewis The PR is ready for review. Could you please have a look as per your availability?

gmlewis reviewed (changes requested):
I don't have time for a full review right now, but at a quick glance, it looks like you are defining a whole bunch of new structs... possibly more than necessary.  We try to reuse structs when the meaning is similar if possible.

It also looks like you are redefining `ListOptions` and/or `ListCursorOptions` when they could be embedded into other options... please look around this repo for examples on how to do this.

Third, all new structs must have complete sentences ending in a period for the auto-generated godocs to be formatted correctly.  Please fix this.

Additionally, it looks like there are linting and/or test errors that need fixing. Please address.

anishrajan25 commented:
@gmlewis I have analysed the existing structs and prepared a document capturing the possibilities of embedding some structs to prevent redefining some fields.

Could you please have a look at [this document](https://docs.google.com/document/d/10ezC0VHeUWk3WBNZJjOlOld97XOLeRQpNA34Xsz0BFM/edit?usp=sharing) and add comments wherever necessary to provide your inputs?

I will fix the comments to end with period, linting, and test errors in the next commit along with updating the struct definitions.

gmlewis commented:
I'll be back on 9/18 when I can take a look at this. Sorry for the delay and inconvenience.

anishrajan25 commented:
No worries. You can have a look whenever you are available ✌🏼

gmlewis commented:
@anishrajan25 - I'm back, but none of the comments I left above:
https://github.com/google/go-github/pull/2902#pullrequestreview-1599004532
have been addressed.  Please fix these issues before we can continue.

anishrajan25 commented:
Hi @gmlewis , apologies for the delay in response.

I have created [this document](https://docs.google.com/document/d/10ezC0VHeUWk3WBNZJjOlOld97XOLeRQpNA34Xsz0BFM/edit?usp=sharing), capturing the possibilities of embedding some structs to prevent redefining some fields. 
Could you please have a look and add comments wherever necessary to provide your input?

anishrajan25 commented:
I have updated the existing structs and reused them instead of creating new ones wherever possible. I have fixed the comments as well. @gmlewis Could you please have a look?

gmlewis commented on github/security_advisories.go:
```diff
@@ -12,6 +12,50 @@ import (
+	Maintain *bool `json:"maintain,omitempty"`
+}
+
+// ListRepositorySecurityAdvisoriesOptions specifies the optional parameters to lists the repository security advisories.
```
```suggestion
// ListRepositorySecurityAdvisoriesOptions specifies the optional parameters to list the repository security advisories.
```

gmlewis commented on github/security_advisories.go:
```diff
@@ -35,3 +79,51 @@ func (s *SecurityAdvisoriesService) RequestCVE(ctx context.Context, owner, repo,
+
+// ListRepositorySecurityAdvisoriesForOrg lists the repository security advisories for an organization.
+//
+// Github API docs: https://docs.github.com/en/rest/security-advisories/repository-advisories?apiVersion=2022-11-28#list-repository-security-advisories-for-an-organization
```
```suggestion
// GitHub API docs: https://docs.github.com/en/rest/security-advisories/repository-advisories?apiVersion=2022-11-28#list-repository-security-advisories-for-an-organization
```

gmlewis commented on github/security_advisories.go:
```diff
@@ -35,3 +79,51 @@ func (s *SecurityAdvisoriesService) RequestCVE(ctx context.Context, owner, repo,
+
+// ListRepositorySecurityAdvisories lists the security advisories in a repository.
+//
+// Github API docs: https://docs.github.com/en/enterprise-cloud@latest/rest/security-advisories/repository-advisories?apiVersion=2022-11-28#list-repository-security-advisories
```
```suggestion
// GitHub API docs: https://docs.github.com/en/enterprise-cloud@latest/rest/security-advisories/repository-advisories?apiVersion=2022-11-28#list-repository-security-advisories
```

gmlewis commented on github/security_advisories.go:
```diff
@@ -35,3 +79,51 @@ func (s *SecurityAdvisoriesService) RequestCVE(ctx context.Context, owner, repo,
+// ListRepositorySecurityAdvisories lists the security advisories in a repository.
+//
+// Github API docs: https://docs.github.com/en/enterprise-cloud@latest/rest/security-advisories/repository-advisories?apiVersion=2022-11-28#list-repository-security-advisories
+func (s *SecurityAdvisoriesService) ListRepositorySecurityAdvisories(ctx context.Context, owner string, repo string, opt *ListRepositorySecurityAdvisoriesOptions) ([]*SecurityAdvisory, *Response, error) {
```
```suggestion
func (s *SecurityAdvisoriesService) ListRepositorySecurityAdvisories(ctx context.Context, owner, repo string, opt *ListRepositorySecurityAdvisoriesOptions) ([]*SecurityAdvisory, *Response, error) {
```

gmlewis commented on github/security_advisories_test.go:
```suggestion
	} else if !strings.Contains(err.Error(), "json: cannot unmarshal number into Go struct field SecurityAdvisory.ghsa_id of type string") {
```

gmlewis commented:
Please check that test pass locally with `go test ./...` before pushing.

anishrajan25 commented:
There are some failing test cases. I will update once the test cases are fixed.

anishrajan25 commented:
@gmlewis I have fixed the test cases as well. All the tests are passing on my local

gmlewis commented:
Please run `go generate ./...` as described in CONTRIBUTING.md and push the results to this PR.

gmlewis reviewed (approved):
Thank you, @anishrajan25 !
LGTM.

Awaiting second LGTM+Approval from any other contributor to the repo before merging.

gmlewis commented:
Thank you, @valbeat !
Merging.
